RationaleMCP/0031/concept.md

```diff
@@ -0,0 +1,135 @@
+
+## Design Rules 
+
+There are simple general rules for the design of Flat Modelica:
+
+* The syntax is a subset of Modelica syntax with some extensions.
+* Everything which has direct influence on the simulation results in an ideal mathematical sense has a special syntactical construct in the language.
+* Information which has no direct influence to the result are handeled uniformly by one syntactical construct called annotation. For example, this applies to
+   * prefixes
+   * attributes of variables 
+   * information lost in preceding transformation steps,
+   * information to control further transformation steps,
+   * information to support the numerical solution of the equation system
+   * source code location information, 
+   * other data for special use cases, e.g. packaging into a FMU, inputs and outputs of the whole model.
+* Description strings (also called syntactic comments) are an exception. These are optionally copied as such.
+* An external tool does not need any knowledge of special Modelica semantics, e.g. handling start attributes, behavior of relations and built-in operators for events,etc..
+
+## Details
+
+* All higher level syntactical and structural contructs of Modelica are removed, such as:
+   * imports
+   * inheritance
+   * redeclarations
+   * nested class definitions (There is no class definition inside another class definition)
+   * packages, [expandable] connectors, models (only one model remains), blocks, operator definitions
+   * functional inputs of functions
+   * conditional declarations
+   * modifications (therefore no merging of them)
+   * connect statements
+   * loops inside equation sections
+   * `if`/`then`/`else` in equation sections (atleast with different sized branches)
+   * `when` clauses
+   * stream operators: `actualStream(...)`, `inStream(...)`
+   * all prefixes, except `input` and `output`
+* __TODO__ synchronous Modelica extensions, including state machines, possibly introduction of partitions
+
+### Names
+
+* Names are defined as the `IDENT` token in Modelica
+* The hierarchy of the Modelica model is encoded into the names of the variables and types by separation of the levels by a dot.
+* Names from Modelica are generally enclosed in single quotation marks, whether or not they contain a dot.  
+* Generated helper variables do not need single quotes. 
+* __TODO__: array components, combining names for helper variables
+
+### Types
+
+* Following types exist:
+   * predefined types: `Real`, `Integer`, `Boolean`, `String`, `ExternalObject`,
+   * record types
+   * functions
+   * simple types.
+   
+* Properties of record definitions are:
+   * The components are defined as variable definitions. Declaration equations are not allowed.
+   * There are no equations or algorithms.
+   * record definitions can not nest.
+   * There is no automatism to generate record constructors anymore. They have to be made explicit as functions.
```

Comment:
Please clarify.  I guess we still need a way to construct values of record types, and that we need a primitive way of doing this that can't be expressed as a user function not relying on a record constructor for its implementation.

Reply:
Why not:
```
record R
  Real x;
  Integer i;
end R;
function RecordConstructorForR
input Real x;
input Real i;
output R r;
algorithm
   r.x=x;
   r.i=i;
end RecordConstructorForR;
```

Reply:
Sure, this looks like a possible way to do it.

Then, my second concern here is that explicitly calling a record constructor is much more easily turned into efficient code (avoiding the need to backwards engineer a function to realize which ones are just record constructors in disguise).  Not having to invent names for the record constructors will also enhance readability of the Flat Modelica code.  Lastly, being forced to create all these record constructor functions seems like unnecessary bloating of the Flat Modelica code.

Would it really be that much of a gain to not have record constructors?

Reply:
If you read in the Modelica specification the section about record constructors, the you see a lot of conditions when an argument should be present or not. This is another example of complited Modelica semantics which should be removed. Especially. for this the information "do not allow modification" is used, which makes no sense in the context of Flat Modelica. So I want to get rid of it.

Reply:
You are right in that the record constructors in Modelica are a little bit complicated, and that it would be nice if Flat Modelica was simpler.  However, I'm just not convinced that it will be so complicated that it can't be kept once we have seen exactly what to do with constants and final members in records (I think you may be right in that there will be no need to keep `final` in Flat Modelica, but I must say I don't understand enough how constant record members are used in Modelica in order to see what role they would play in Flat Modelica except for storing non-scalarized model variables).  If we also drop support for named arguments in general, this will also simplify record constructors even more.

Perhaps record constructors is another example that we should keep in a first version of Flat Modelica, and then see if we can/want to get rid of in a later version?